crates/rb-allocator/src/lib.rs

```diff
@@ -0,0 +1,34 @@
+//! Allocator which reports mem usage to the Ruby GC
+//!
+//! This crate exports, one type, `RbAllocator`, which implements the `GlobalAlloc` trait. It is a
+//! simple wrapper over the system allocator which reports memory usage to Ruby using
+//! `rb_gc_adjust_memory_usage`
+
+use rb_sys::{rb_gc_adjust_memory_usage, ssize_t};
+use std::alloc::{GlobalAlloc, Layout, System};
+
+pub struct RbAllocator;
+
+unsafe impl GlobalAlloc for RbAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let ret = System.alloc(layout);
+        if !ret.is_null() {
+            rb_gc_adjust_memory_usage(layout.size() as ssize_t)
```

Comment:
[The docs][gc.h] aren't super clear, and [the code][gc.c] is a bit beyond me but it _looks_ like this can trigger GC, such that it would make sense to call it _before_ the allocator, so that the CG can make room for the allocation.

How to account for the possibility of the allocation failing I'm not sure, `rb_gc_adjust_memory_usage` back down with a negative size I guess?

[gc.h]: https://github.com/ruby/ruby/blob/5c843a1a6e24aeabb3497065a362caf7b3e2d3b1/include/ruby/internal/intern/gc.h#L376-L388
[gc.c]: https://github.com/ruby/ruby/blob/5c843a1a6e24aeabb3497065a362caf7b3e2d3b1/gc.c#L11625

Reply:
`ruby_xmalloc` actually does do this actual the allocation, although I'm not sure why.  It does seem like it would make more sense for GC to be triggered before the allocation.